### PR TITLE
Added thread local buffers to the data reader class so that it can

### PR DESCRIPTION
**Pull request by bvanessen**
decode image data into a temporary space that is declared once.  Also
updated the load_file function to load data with a single read rather
than using an interator.

### CODE REVIEW

File: include/lbann/utils/file_utils.hpp [hunk 1]
```diff
@@ -65,7 +65,7 @@ bool create_dir(const std::string output_dir);
 
 
 /// Load a file into a buffer
-template <typename T = std::vector<unsigned char> >
+template <typename T = std::vector<char> >
```
Comment (timmoon10) [inline include/lbann/utils/file_utils.hpp]: Do we need this template?

File: src/data_readers/data_reader.cpp [hunk 2]
```diff
@@ -58,6 +58,13 @@ void generic_data_reader::setup() {
   set_initial_position();
 
   shuffle_indices();
+
+  m_thread_buffer.resize(omp_get_max_threads(), std::vector<char>());
+#pragma omp parallel
```
Comment (timmoon10) [inline src/data_readers/data_reader.cpp]: It seems simpler to loop through the buffers rather than launching an OpenMP section.

Comment (bvanessen) [reply src/data_readers/data_reader.cpp]: Good point.

File: include/lbann/utils/file_utils.hpp (hunk 1 above)
Comment (bvanessen) [reply include/lbann/utils/file_utils.hpp]: No, the previous version of code made use of the template, but we don't need it.

**Pull request merged**